### PR TITLE
Close #131

### LINKED ISSUE #131: Mobile navigation: Should highlight current section
### Please leave a comment if you are going to work on this issue, thank you.

Right now, in the mobile navigation there is no indication about which section I am viewing. Maybe better to add an indicator (such as making the text bold) for current item.

image

### PR DESCRIPTION
**Pull request by nontachaiwebdev**
This is active version.I think this is not a right way to solve this problem.I think the right way is check handle scroll event and get now section position.Maybe i have a something wrong about you want.

Close #131

### CODE REVIEW

File: components/LevitatingLink.js [hunk 1]
```diff
@@ -1,7 +1,8 @@
-export default function LevitatingLink ({ href, children }) {
+export default function LevitatingLink ({ href, children, handleClick, active }) {
```
Comment (dtinth) [inline components/LevitatingLink.js]: I think adding this functionality shouldn’t need to cause this component to change. We should be detecting the active menu item based on scroll.

**Review by dtinth — CHANGES_REQUESTED** [review]
@nontachaiwebdev Thanks for your PR.

You were right that the correct solution would check the scrolling position to figure out which menu item should be highlighted.

I recommend adding that logic into `MobileNavigation#handleScroll` function.

File: components/LevitatingLink.js (hunk 1 above)
Comment (nontachaiwebdev) [reply components/LevitatingLink.js]: Thank for your advice.Can in use getElementByID instead Refs.How do you think about this point?

Comment (dtinth) [reply components/LevitatingLink.js]: @nontachaiwebdev Yep, you may use `document.getElementById` and all the [data](https://github.com/reactbkk/2.0.0/blob/c1e5fcc6007fa7dafb68eeda3431574ce3b22334/pages/index.js#L15-L22) you need is already given in `this.props.navs` ^^

**Comment by nontachaiwebdev** [comment]
Waiting for you review my code kub.
Thank you

File: components/LevitatingLink.js [hunk 2]
```diff
@@ -1,7 +1,8 @@
-export default function LevitatingLink ({ href, children }) {
+export default function LevitatingLink ({ href, children, active }) {
```
Comment (dtinth) [inline components/LevitatingLink.js]: A LevitatingLink should not take an `active` prop, because the `active` prop is for the navigation. A general UI component like LevitatingLink does not need to know it.

You can add an `active` class outside this component, e.g., inside `NavigationLink`.

File: components/MobileNavigation.js [hunk 3]
```diff
@@ -20,23 +20,53 @@ export default class MobileNavigation extends Component {
...
+    scrollPosition: null
   }
 
   componentDidMount () {
+    const { navs } = this.props
     window.addEventListener('scroll', this.handleScroll)
+    const scrollPosition = navs.reduce((obj, nav, index) => {
+      const navScroll = {
+        name: nav.label,
+        scroll: document.getElementById(nav.label.toLowerCase()).getBoundingClientRect().top + window.scrollY
```
Comment (dtinth) [inline components/MobileNavigation.js]: Don’t precalculate the scroll position, because the page may be resized. I think it is fine to calculate it on-the-fly (during scrolling).

Comment (dtinth) [inline components/MobileNavigation.js]: `document.getElementById(nav.label.toLowerCase())`

Do not depend on the label, because scrolling is based on `href`. Instead, use `document.getElementById(nav.href.slice(1))`.

Comment (dtinth) [reply components/MobileNavigation.js]: Instead of storing the positions, you can store the elements instead. Then we can call the array `navigationTargets`.

File: components/MobileNavigation.js [hunk 4]
```diff
@@ -20,23 +20,53 @@ export default class MobileNavigation extends Component {
...
+      }
+      obj[index] = navScroll
+      return obj
+    }, [])
+    this.setState({
+      scrollPosition: scrollPosition
```
Comment (dtinth) [inline components/MobileNavigation.js]: Since `scrollPosition` is not used in `render()`, there is no need to store it inside state.

Instead, you can just store it as an instance attribute: `this.navigationTargets = navigationTargets`.

File: components/MobileNavigation.js [hunk 5]
```diff
@@ -20,23 +20,53 @@ export default class MobileNavigation extends Component {
...
+    nowSelectMenu: null,
+    scrollPosition: null
   }
 
   componentDidMount () {
+    const { navs } = this.props
     window.addEventListener('scroll', this.handleScroll)
+    const scrollPosition = navs.reduce((obj, nav, index) => {
+      const navScroll = {
+        name: nav.label,
```
Comment (dtinth) [inline components/MobileNavigation.js]: Don’t depend on the label since it is for displaying purpose only.

Instead, use `nav.href` here, since it is used to ‘ref’er to sections on the page.

File: components/MobileNavigation.js [hunk 6]
```diff
@@ -20,23 +20,53 @@ export default class MobileNavigation extends Component {
...
+    })
   }
 
   componentWillUnmount () {
     window.removeEventListener('scroll', this.handleScroll)
   }
 
   handleScroll = (e) => {
+    const _docHeight = (document.height !== undefined) ? document.height : document.body.offsetHeight
+    let result
```
Comment (dtinth) [inline components/MobileNavigation.js]: You can avoid a mutable variable here. Try solving the problem using only `const` (no `let` and `var`).

**Review by dtinth — CHANGES_REQUESTED** [review]
Here’s your review krub 😄

**Comment by dtinth** [comment]
Please don't forget to add an issue tag to this pull request.

Check out this comment: https://github.com/reactbkk/2.0.0/pull/153#issuecomment-304457716

**Comment by nontachaiwebdev** [comment]
thank you alot for you review.You review that very improve my skill.

File: components/MobileNavigation.js (hunk 6 above)
Comment (nontachaiwebdev) [reply components/MobileNavigation.js]: Hi am almost complete all thing your review but i have one question.Do you thing about switch result variable from let to const ? I think Because result variable want able to change value.If i declare const in condition scope.i will can't call result const outside condition scope to set new state.Any idea. Thankyou.

Comment (dtinth) [reply components/MobileNavigation.js]: Thanks for the question. What I mean is that you can write code in a way that does not require changing a variable. For example, you can do:

```js
    const selectedMenu = this.findSelectedMenuFromScroll()
    if (this.state.nowSelectMenu !== selectedMenu) {
      this.setState({ nowSelectMenu: selectedMenu })
    }
```

Then you can create a separate method like this:

```js
  findSelectedMenuFromScroll () {
    ...
    if (...) {
      return ...
    } else {
      return ...
    }
  }
```

By factoring out functions, you can eliminate variables that can change.

File: components/LevitatingLink.js [hunk 7]
```diff
@@ -1,7 +1,7 @@
 export default function LevitatingLink ({ href, children }) {
   return (
-    <a href={href}>
-      <span className='text'>{children}</span>
+    <a href={href} >
+      <span>{children}</span>
```
Comment (dtinth) [inline components/LevitatingLink.js]: Please revert the changes in this file because it is breaking the link animation. (The text does not levitate when hover)

File: components/MobileNavigation.js [hunk 8]
```diff
@@ -77,8 +106,7 @@ export default class MobileNavigation extends Component {
               <div className='nav-header' key='nav-header'>React Bangkok</div>
             </NavigationLink>
           }
-          { expandMobileNav && navs.map(nav =>
-            <NavigationLink href={nav.href} key={nav.href} disabled={nav.disabled}>{nav.label}</NavigationLink>)
+          { expandMobileNav && navs.map(nav => (<NavigationLink href={nav.href} key={nav.href} active={nowSelectMenu === nav.label} disabled={nav.disabled}>{nav.label}</NavigationLink>))
```
Comment (dtinth) [inline components/MobileNavigation.js]: This line is too long. Please break it down:

```js
           {expandMobileNav && navs.map(nav => (
             <NavigationLink
               href={nav.href}
               key={nav.href}
               active={nowSelectMenu === nav.label}
               disabled={nav.disabled}
             >
               {nav.label}
             </NavigationLink>
           ))}
```

File: components/MobileNavigation.js [hunk 9]
```diff
@@ -20,11 +19,22 @@ export default class MobileNavigation extends Component {
...
     window.addEventListener('scroll', this.handleScroll)
+    const target = navs.reduce((obj, nav, index) => {
+      const navTarget = {
+        name: nav.label,
+        target: document.getElementById(nav.href.slice(1))
+      }
+      obj[index] = navTarget
+      return obj
+    }, [])
+    this.navigationTarget = target
```
Comment (dtinth) [inline components/MobileNavigation.js]: Since there are multiple targets, please name this `navigationTargets`.

File: components/MobileNavigation.js [hunk 10]
```diff
@@ -20,11 +19,22 @@ export default class MobileNavigation extends Component {
...
+    const { navs } = this.props
     window.addEventListener('scroll', this.handleScroll)
+    const target = navs.reduce((obj, nav, index) => {
+      const navTarget = {
+        name: nav.label,
+        target: document.getElementById(nav.href.slice(1))
+      }
+      obj[index] = navTarget
+      return obj
+    }, [])
```
Comment (dtinth) [inline components/MobileNavigation.js]: Since you are reducing into another array keeping the order, this is the same as mapping. Better do this:

```js
    const navigationTargets = navs.map((nav) => {
      return {
        name: nav.label,
        target: document.getElementById(nav.href.slice(1))
      }
    })
```

File: components/MobileNavigation.js [hunk 11]
```diff
@@ -20,11 +19,22 @@ export default class MobileNavigation extends Component {
...
     showMobileNav: false,
-    expandMobileNav: false
+    expandMobileNav: false,
+    nowSelectMenu: null
   }
 
   componentDidMount () {
+    const { navs } = this.props
     window.addEventListener('scroll', this.handleScroll)
+    const target = navs.reduce((obj, nav, index) => {
```
Comment (dtinth) [inline components/MobileNavigation.js]: Let’s name this variable `navigationTargets` because it hold multiple items.

File: components/MobileNavigation.js [hunk 12]
```diff
@@ -20,11 +19,22 @@ export default class MobileNavigation extends Component {
 
   state = {
     showMobileNav: false,
-    expandMobileNav: false
+    expandMobileNav: false,
+    nowSelectMenu: null
```
Comment (dtinth) [inline components/MobileNavigation.js]: Let’s name this `activeMenu` for more correct grammar.

File: components/MobileNavigation.js [hunk 13]
```diff
@@ -37,6 +47,26 @@ export default class MobileNavigation extends Component {
     } else {
       if (this.state.showMobileNav) this.hideMobileNav()
     }
+    const result = this.findScrollSection()
+    this.setState({
```
Comment (dtinth) [inline components/MobileNavigation.js]: If the result doesn’t change, there is no need to update the state (which triggers a rerender).

Better guard it:

```js
    const result = this.findScrollSection()
    if (this.state.activeMenu !== result) {
      this.setState({ nowSelectMenu: result })
    }
```

File: components/MobileNavigation.js [hunk 14]
```diff
@@ -37,6 +47,26 @@ export default class MobileNavigation extends Component {
...
+    const _docHeight = (document.height !== undefined) ? document.height : document.body.offsetHeight
+    if ((_docHeight - window.innerHeight) === window.scrollY) {
+      return navigationTarget[navigationTarget.length - 1].name
+    } else {
+      return navigationTarget.reduce((result, item) => {
+        const targetScroll = item.target.getBoundingClientRect().top + window.scrollY
+        if (window.scrollY >= (targetScroll - 50)) {
+          result = item.name
+        }
+        return result
```
Comment (dtinth) [inline components/MobileNavigation.js]: Lines 64~67 can be shortened into:

```js
        return window.scrollY >= targetScroll - 50 ? item.name : result
```

So that we can avoid re-assigning to the `result` variable.

File: components/MobileNavigation.js [hunk 15]
```diff
@@ -148,10 +176,10 @@ export default class MobileNavigation extends Component {
   }
 }
 
-function NavigationLink ({ href, disabled, children }) {
+function NavigationLink ({ href, disabled, children, active }) {
   return (
-    <div className={disabled ? 'disabled' : ''}>
-      <LevitatingLink href={href}>{children}</LevitatingLink>
+    <div className={active ? 'disabled' : ''}>
```
Comment (dtinth) [inline components/MobileNavigation.js]: The class name should represent what it’s doing!

Now `disabled` doesn’t work anymore...

The correct class name should look like this:

| disabled | active | result className |
| --- | --- | --- |
| `false` | `false` | `''` |
| `false` | `true` | `'active'` |
| `true` | `false` | `'disabled'` |
| `true` | `true` | `'disabled active'` |

You can do it like this:

```js
  const className = [
    disabled && 'disabled',
    active && 'active'
  ].filter(name => name).join(' ')

  return (
    <div className={className}>
      ...
```

File: components/MobileNavigation.js [hunk 16]
```diff
@@ -164,8 +192,9 @@ function NavigationLink ({ href, disabled, children }) {
           opacity: 0.5;
         }
         .disabled {
-          opacity: 0.25;
-          pointer-events: none;
+          /* opacity: 0.25;
+          pointer-events: none; */
```
Comment (dtinth) [inline components/MobileNavigation.js]: Don’t hack to change disabled into active state. Please create a new class `.active` to represent the active state.

**Review by dtinth — CHANGES_REQUESTED** [review]
Here is your review. :)

Almost there~

**Comment by dtinth** [comment]
Also, please update the pull request **description** (not the title) to include an issue tag.

**Review by dtinth — APPROVED** [review]
Thanks for your PR. This looks good now~
